### ISSUE TITLE
`torch.utils.checkpoint` fail to propagate `backward` to all parameters

### ISSUE THREAD

**Issue by CharlesJQuarra** [pytorch/pytorch#11944]
## Issue description

Discovered while attempting to write a linear unit that supports splitting the parameter space in a grid of gradient checkpoint nodes. The issue right now is that when there is more than one segment, the `backward()` only updates the gradient for the last parameter. More details in the code example:

## Code example
```
import torch
import torch.nn as nn
import torch.utils.checkpoint as checkpoint

def get_segments(total, max_length):
  segments = (total // max_length)
  seg_array = segments*[max_length]
  remainder = total % max_length
  if remainder > 0:
    seg_array += [remainder]
  return seg_array

class GradCheckpoint_Linear(nn.Module):
  def __init__(self, in_features, out_features, cpc_specs={}):
    super(GradCheckpoint_Linear, self).__init__()
    self.in_features = in_features
    self.out_features = out_features

    if 'in_max_segment' in cpc_specs:
      in_max_segment = cpc_specs['in_max_segment']
    else:
      in_max_segment = in_features
    self.in_segment_lengths = get_segments(in_features, in_max_segment)
    if 'out_max_segment' in cpc_specs:
      out_max_segment = cpc_specs['out_max_segment']
    else:
      out_max_segment = out_features
    if 'initializer' in cpc_specs:
      self.initializer = cpc_specs['initializer']
    else:
      def get_init(w, h):
        return torch.randn(w,h)
      self.initializer = get_init
    self.out_segment_lengths = get_segments(out_features, out_max_segment)

    print("in_segment_lengths: {0}".format(self.in_segment_lengths))
    print("out_segment_lengths: {0}".format(self.out_segment_lengths))
    weight_parameters_ = []
    bias_parameters_ = []

    self.array_to_weight_param = -torch.ones(len(self.in_segment_lengths), len(self.out_segment_lengths), dtype=torch.int32)
    for in_idx, in_s_length in enumerate(self.in_segment_lengths):
      for out_idx, out_s_length in enumerate(self.out_segment_lengths):
        param = nn.Parameter( self.initializer(out_s_length,in_s_length) )
        self.array_to_weight_param[in_idx,out_idx]=len(weight_parameters_)
        weight_parameters_.append( param )

    self.weight_parameters = nn.ParameterList( weight_parameters_ )

    for out_s_length in self.out_segment_lengths:
      bias_parameters_.append( nn.Parameter( self.initializer(1, out_s_length).view(out_s_length) ) )
    self.bias_parameters = nn.ParameterList( bias_parameters_ )

  def reset_parameters(self):
    pass

  def forward(self, inp):
    unit_outs = []
    for out_idx, out_s_length in enumerate(self.out_segment_lengths):
      bias_param = self.bias_parameters[out_idx]
      in_offset = 0
      weight_outs = []
      for in_idx, in_s_length in enumerate(self.in_segment_lengths):
        weight_param = self.weight_parameters[ self.array_to_weight_param[in_idx, out_idx] ]
        def fwd_unit_segment(inp_):
          return torch.mv(weight_param, inp_)
        weight_out = checkpoint.checkpoint( fwd_unit_segment , inp[in_offset:in_offset+in_s_length] )
        in_offset += in_s_length
        weight_outs.append(weight_out)
      unit_outs.append( bias_param + sum(weight_outs) )
    result = torch.cat(unit_outs)
    return result

def init_(w,h):
  return torch.ones(w,h)

u_original = GradCheckpoint_Linear(6, 2, cpc_specs={'initializer': init_})
u_split = GradCheckpoint_Linear(6, 2, cpc_specs={'in_max_segment': 2, 'out_max_segment': 1, 'initializer': init_})

inp_0_orig = torch.ones(6, requires_grad=True)
inp_0_split = torch.ones(6, requires_grad=True)

u_original(inp_0_orig).sum().backward()
u_original.weight_parameters[0].grad  #<--- looks good

def top_node(inp_):
  return u_split(inp_)

v = checkpoint.checkpoint(top_node, inp_0_split )
v.sum().backward()

u_split.weight_parameters[0].grad #<--- grad is None
u_split.weight_parameters[5].grad #<--- only grad parameter that has values is the last one?

```

## System Info
Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 9.0.176
GPU models and configuration:
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.4
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip3] numpy (1.14.3)
[pip3] torch (0.4.0)
[pip3] torchfile (0.1.0)
[pip3] torchnet (0.0.2)
[conda] cuda91                    1.0                  h4c16780_0    pytorch
[conda] pytorch                   0.4.0           py36_cuda9.1.85_cudnn7.1.2_1  [cuda91]  pytorch
[conda] torchvision               0.2.0            py36h17b6947_1    pytorch

**Comment by t-vi** [comment]
It's not a bug in the checkpointing, but it is the dynamic scoping in Python:
During the backward,
```
 def fwd_unit_segment(inp_):
          return torch.mv(weight_param, inp_)
```
and that will grab the current(!) (i.e. last processed) weight_param.
If you change that to
```
        def fwd_unit_segment(inp_, weight_param=weight_param):
          return torch.mv(weight_param, inp_)
```
(the not so pretty way of binding the weight_param at evaluation time), you get what you expect.

Please accept my apologies for not noticing earlier.

**Comment by CharlesJQuarra** [comment]
confim that this solves the issue. Thank you @t-vi !
